envbuilder.go

```diff
@@ -58,32 +58,65 @@ var ErrNoFallbackImage = errors.New("no fallback image has been specified")
 // DockerConfig represents the Docker configuration file.
 type DockerConfig configfile.ConfigFile
 
+type runtimeDataStore struct {
+	// Runtime data.
+	Image            bool                          `json:"-"`
+	Built            bool                          `json:"-"`
+	SkippedRebuild   bool                          `json:"-"`
+	Scripts          devcontainer.LifecycleScripts `json:"-"`
+	ImageEnv         []string                      `json:"-"`
+	ContainerEnv     map[string]string             `json:"-"`
+	RemoteEnv        map[string]string             `json:"-"`
+	DevcontainerPath string                        `json:"-"`
+
+	// Data stored in the magic image file.
+	ContainerUser string `json:"container_user"`
+}
+
+type execArgsInfo struct {
+	InitCommand string
+	InitArgs    []string
+	UserInfo    userInfo
+	Environ     []string
+}
+
 // Run runs the envbuilder.
 // Logger is the logf to use for all operations.
 // Filesystem is the filesystem to use for all operations.
 // Defaults to the host filesystem.
 func Run(ctx context.Context, opts options.Options) error {
-	defer options.UnsetEnv()
-	if opts.GetCachedImage {
-		return fmt.Errorf("developer error: use RunCacheProbe instead")
+	var args execArgsInfo
+	// Run in a separate function to ensure all defers run before we
+	// setuid or exec.
+	err := run(ctx, opts, &args)
```

Comment:
Review: Even though we've fixed defers here, there's still `cmd/envbuilder` where e.g. `closeLogs` won't fire since we execve(2) after this.

Reply:
https://github.com/coder/envbuilder/issues/364